### PR TITLE
[new release] goblint (2.2.1)

### PR DESCRIPTION
**Pull request by sim642**
Static analysis framework for C

- Project page: <a href="https://goblint.in.tum.de">https://goblint.in.tum.de</a>
- Documentation: <a href="https://goblint.readthedocs.io/en/latest/">https://goblint.readthedocs.io/en/latest/</a>

##### CHANGES:

* Bump batteries lower bound to 3.5.0.
* Fix flaky dead code elimination transformation test.

### CODE REVIEW

**Comment by sim642** [comment]
All the CI failures aren't really related nor fixable here.

**Comment by tmcgilchrist** [comment]
The macOS failure is curious, batteries not finding a real gcc? Perhaps it is related to the FreeBSD failure which is caused by conf-gcc not having an entry for FreeBSD. 
Add FreeBSD for conf-gcc https://github.com/ocaml/opam-repository/pull/24429

**Comment by sim642** [comment]
The "couldn't find real gcc" error is from goblint-cil, not goblint. I'm surprised to see this again because in https://github.com/ocaml/opam-repository/pull/21975#issuecomment-1213132109 I said:

> With conf-gcc it's depext forces homebrew to install a real gcc.

Depext also shows gcc being installed via homebrew, so I'm not sure what the issue is.

File: packages/lintcstubs/lintcstubs.0.4.7/opam [hunk 1]
```diff
@@ -11,7 +11,7 @@ depends: [
   "dune" {>= "3.0"}
   "ocaml" {>= "4.13"}
   "lintcstubs-gen" {= version}
-  "goblint" {>= "2.1.0"}
+  "goblint" {>= "2.1.0" & < "2.2.0"}
```
Comment (sim642) [inline packages/lintcstubs/lintcstubs.0.4.7/opam:14]: @edwintorok Goblint's library interface hasn't been designed for external use and regularly breaks between releases, so you'll probably want to adapt lintcstubs. Previous Goblint releases are incompatible with OCaml 5.0, which also means lintcstubs is, which is unfortunate (OCaml 5 C stubs have to be analyzed under OCaml 4.14 instead).

Comment (edwintorok) [reply packages/lintcstubs/lintcstubs.0.4.7/opam:14]: No problem, I'll add some tighter bounds in my repo, thanks for fixing it up. In general I intend to track the latest release of goblint, I think master of goblint already works with OCaml 5.

Comment (edwintorok) [reply packages/lintcstubs/lintcstubs.0.4.7/opam:14]: https://github.com/edwintorok/lintcstubs/tree/goblint2.2.1-compat I've done an experiment to upgrade to latest goblint on this branch, there are some new warnings from goblint, but seems to work so far and the diff is fairly small: https://github.com/edwintorok/lintcstubs/commit/ce8150a60e7a013c33a93090b80596d4a1d0b769

**Comment by raphael-proust** [comment]
On the macos CI failure I see the following in the logs.

First, it installs the `conf-gcc` package and that seems to happen alright.

```
- ==> Installing gcc
- ==> Pouring gcc--13.2.0.ventura.bottle.1.tar.gz
- No entry for terminal type "unknown";
- using dumb terminal settings.
- 🍺  /usr/local/Cellar/gcc/13.2.0: 1,590 files, 421.7MB
```

Then as part of installing `goblint-cil`, it tries to discover the local available the gcc by checking different versions. It starts with just plain `gcc`:

```
# (cd _build/default/bin && ./realGccConfigure.exe)
# run: gcc --version
# -> process exited with code 0
# -> stdout:
#  | Apple clang version 14.0.3 (clang-1403.0.22.14.1)
#  | Target: x86_64-apple-darwin22.5.0
#  | Thread model: posix
#  | InstalledDir: /Library/Developer/CommandLineTools/usr/bin
# -> stderr:
```

And then tries to go down from version to version:

```
# run: gcc-12 --version
# -> process exited with code 127
# -> stdout:
# -> stderr:
#  | sh: gcc-12: command not found
# run: gcc-11 --version
# -> process exited with code 127
# -> stdout:
# -> stderr:
#  | sh: gcc-11: command not found
# run: gcc-10 --version
# -> process exited with code 127
# -> stdout:
# -> stderr:
#  | sh: gcc-10: command not found
# run: gcc-9 --version
# -> process exited with code 127
# -> stdout:
# -> stderr:
#  | sh: gcc-9: command not found
# run: gcc-8 --version
# -> process exited with code 127
# -> stdout:
# -> stderr:
#  | sh: gcc-8: command not found
# run: gcc-7 --version
# -> process exited with code 127
# -> stdout:
# -> stderr:
#  | sh: gcc-7: command not found
# Fatal error: exception Failure("couldn't find real gcc")
```

As you can see, it misses the actual version that `goblint-cil` is actually installed at.

I'm a bit confused because AFAICT it should try versions 16 and lower. I think the issue is that `goblint-cil<2.02` should be marked as incompatible with `macos`? or `conf-gcc`? or something like that?

**Comment by raphael-proust** [comment]
@sim642 sorry to be using your PR to investigate and fix some not-totally-related issues, just trying to take the opportunity to improve the general health of the opam-repository.

**Comment by edwintorok** [comment]
Latest goblint-cil builds just  fine on my Mac OS (with gcc-13). Earlier 'goblint-cil' indeed doesn't build *on the latest version* of 'gcc' that homebrew installs. If you install a different version of 'gcc' it might work, but I don' think you can express that in 'conf-gcc'. To make the CI happier you could indeed mark older version of 'goblint-cil' as not supported, but please only do that on the CI (there are some x-ci-accept-failures tags), it may be possible for someone using an actual Mac to get it working.

**Comment by sim642** [comment]
> I'm a bit confused because AFAICT it should try versions 16 and lower. I think the issue is that `goblint-cil<2.02` should be marked as incompatible with `macos`? or `conf-gcc`? or something like that?

Ah! That lintcstubs macos job is forced to go with goblint-cil 2.0.1, which I didn't notice before. goblint-cil 2.0.2 extends the range of gcc binaries it looks for such that it works.

Indeed, it works on macos if an older gcc from Homebrew is present. So goblint-cil 2.0.1 isn't outright incompatible with macos but depends on some depext versions on which constraints aren't possible.

**Comment by raphael-proust** [comment]
Is it likely to work if we just add a `patch` file for `goblint-cil` to extend gcc version detection?

We could add this patch to the package. I think that's the solution which keeps the most compatibility. Does anyone know if `goblint-cil` might have incompatibilities with higher versions of gcc?

**Comment by sim642** [comment]
It should be fine. This patch should do it: https://github.com/goblint/cil/commit/28d3e7e7b3373e03b4a2e35ef6f446bd66489743.

**Comment by raphael-proust** [comment]
I made https://github.com/ocaml/opam-repository/pull/24493

**Comment by raphael-proust** [comment]
I think this is ready to merge.

**Comment by dinosaure** [comment]
Thanks!

**Pull request merged**